scalyr_agent/log_processing.py

```diff
@@ -88,45 +88,150 @@
 
 log = scalyr_logging.getLogger(__name__)
 
+class CRIParseError(Exception):
+    def __init__(self, line, message):
+        self.line = line
+        self.message = message
 
-def _parse_cri_log(line):
-    """
-    Parse a log line that uses the K8S Container Runtime Interface (CRI) format
-    https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/kubelet-cri-logging.md
-    """
-    # find the end of the timestamp segment
-    index = line.find(" ")
-    if index < 0:
-        return None, None, None, None
+    def __repr__(self):
+        return "Error parsing line - %s: %s" % self.message, self.line
+
+class CRILogLine(object):
+    __slots__ = ("timestamp", "stream", "tags", "message", "raw_timestamp", "is_partial", "is_full")
+
+    def __init__(self, timestamp, stream, tags, message, raw_timestamp):
+        self.timestamp = timestamp
+        self.stream = stream
+        self.tags = tags
+        self.message = message
+        self.raw_timestamp = raw_timestamp
+
+        self.is_partial = "P" in tags if tags else False
+        self.is_full = "F" in tags if tags else False
+
+    @classmethod
+    def from_raw_line(cls, raw_line):
+        decoded_raw_line = raw_line.decode("utf-8", "replace")
+        timestamp, stream, tags, message, raw_timestamp = cls.__parse_cri_log(
+            decoded_raw_line
+        )
+
+        log_line = cls(timestamp, stream, tags, message, raw_timestamp)
+
+        if not log_line.is_partial and not log_line.is_full:
```

Comment:
Should also verify that is_partial & is_full are not both set

Reply:
I am not sure if I understand. They will be always set to True or False.
This line is to check the validity of the raw_line.

Reply:
I meant it should also check that both is_partial & is_full are not both true, ie:
```
if (not log_line.is_partial and not log_line.is_full) or (log_line.is_partial and log_file.is_full):
   ...
```

Reply:
is_partial and is_full will both be always set int the consturctor:
```
self.is_partial = "P" in tags if tags else False
self.is_full = "F" in tags if tags else False
```